### PR TITLE
Automatic evaluation for integration test.

### PR DESCRIPTION
**Pull request by hooshmandshr**
The following things have been added.

- Function to run evaluations on a select set of data sets.

- Class to generate evaluation plots.

- A set of simple auxiliary functions like, main channel detection, PNR computation and logit transform.

A set of datasets (EJ retinal) alongside gold standard spike trains are required for this to work. It take about 4GB of data. On top of that the script will generate 4 sets of fake data sets that take about 4GB more. I have not added these data sets.

### CODE REVIEW

File: tests/evaluation/retinal_evaluation.py [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+    for data_idx, data_number in enumerate(range(1, k_tot_data + 1)):
+        # Setting up config file for yass.
+        bin_file = 'ej49_data{}.bin'.format(data_number)
+        geom_file = 'ej49_geometry{}.txt'.format(data_number)
+        config['data']['recordings'] = bin_file
+        config['data']['geometry'] = geom_file
+        with open('config_ej49.yaml', 'w') as yaml_file:
+            yaml.dump(config, yaml_file, default_flow_style=False)
+        # TODO(hooshmand): Find the internal call for yass.
+        call(['yass', 'config_ej49.yaml'])
```
Comment (calvinytong) [inline tests/evaluation/retinal_evaluation.py]: @edublancas what is the call for this? It is a little hacky to invoke the cli from a program that already imports the package.

File: tests/evaluation/retinal_evaluation.py [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+        aug_gold_spt, status = stab.save_augment_recording(
+                aug_bin_file, n_batches)
+        np.save('ej49_data{}.aug.npy'.format(data_number), aug_gold_spt)
+        # Setting up config file for yass to run on augmented data.
+        config['data']['recordings'] = aug_bin_file
+        config['data']['geometry'] = geom_file
+        with open('config_ej49.yaml', 'w') as yaml_file:
+            yaml.dump(config, yaml_file, default_flow_style=False)
+        # TODO(hooshmand): Find the internal call for yass.
+        call(['yass', 'config_ej49.yaml'])
```
Comment (calvinytong) [inline tests/evaluation/retinal_evaluation.py]: Same comment as above

File: tests/evaluation/retinal_evaluation.py [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+    stb_plot = EvaluationPlot('EJ Retinal', k_tot_data, eval_type='Stability')
+    acc_plot = EvaluationPlot('EJ Retinal', k_tot_data)
+
+    for data_idx, data_number in enumerate(range(1, k_tot_data + 1)):
+        # Setting up config file for yass.
+        bin_file = 'ej49_data{}.bin'.format(data_number)
+        geom_file = 'ej49_geometry{}.txt'.format(data_number)
+        config['data']['recordings'] = bin_file
+        config['data']['geometry'] = geom_file
+        with open('config_ej49.yaml', 'w') as yaml_file:
```
Comment (calvinytong) [inline tests/evaluation/retinal_evaluation.py]: @edublancas Do we have methods for running yass directly from a dictionary? It seems like an extra step to dump the config settings into a yaml file from code when we can just feed it directly into the algorithm

File: tests/evaluation/retinal_evaluation.py (hunk 1 above)
Comment (edublancas) [reply tests/evaluation/retinal_evaluation.py]: there is a `_run_pipeline` function in the `command_line.py`, however, it is not indented for public use, it can be safely used for now, I should probably create a function with the complete pipeline

Comment (edublancas) [reply tests/evaluation/retinal_evaluation.py]: Also, `yass config.yaml` changed to `yass sort config.yaml`

File: tests/evaluation/retinal_evaluation.py (hunk 3 above)
Comment (edublancas) [reply tests/evaluation/retinal_evaluation.py]: Agree with the dictionary thing, there is currently no way to do it directly from a dictionary. @hooshmandshr can you open an issue about this? this is also related to the next comment, I need to add a function that runs the entire pipeline and also that accepts a dictionary. It shouldn't take much time to be done

File: tests/evaluation/retinal_evaluation.py (hunk 1 above)
Comment (hooshmandshr) [reply tests/evaluation/retinal_evaluation.py]: I just created an issue with running the entire pipeline using an internal function and a config dictionary. I replaced it with _run_pipeline() for now.

Comment (hooshmandshr) [reply tests/evaluation/retinal_evaluation.py]: I agree with your point Calvin, we do need to have some interface function to get spike trains etc.

File: tests/evaluation/retinal_evaluation.py (hunk 2 above)
Comment (hooshmandshr) [reply tests/evaluation/retinal_evaluation.py]: Replaced with a temporary internal call.

File: tests/evaluation/retinal_evaluation.py (hunk 3 above)
Comment (hooshmandshr) [reply tests/evaluation/retinal_evaluation.py]: Agreed. I just opened an issue on this.

File: tests/evaluation/retinal_evaluation.py [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+                    print("No metric found for {} for dataset {}".format(
+                        method, i + 1))
+        self.fig.set_size_inches(12, 4 * self.n_dataset)
+        plt.savefig('{}_{}.pdf'.format(self.data_set_title, self.eval_type))
+
+
+def main(n_batches=6):
+    """Runs the procedure for evaluating yass on retinal data."""
+
+    config_file = open('config_template.yaml', 'r')
```
Comment (calvinytong) [inline tests/evaluation/retinal_evaluation.py]: @hooshmandshr I tried running this, but I couldn't find this file. Is this checked in somewhere?

Comment (hooshmandshr) [reply tests/evaluation/retinal_evaluation.py]: I just uploaded geometry, ground truth and config files that are required.
However, I don't know what to do with the binary files 4 x 1GB that I cannot commit here.

Comment (calvinytong) [reply tests/evaluation/retinal_evaluation.py]: This may be a problem because the integration test needs those datasets to run on push. Usually, i'd recommend something on s3, but I really don't want to set that up. Is there some way we can trim the amount of data we use?

Comment (hooshmandshr) [reply tests/evaluation/retinal_evaluation.py]: I understand your concern,
unfortunately, I'm not sure we can trim to much smaller that this

Comment (edublancas) [reply tests/evaluation/retinal_evaluation.py]: We can run the this nightly in our server with the full dataset and a trimmed version in travis.

Comment (hooshmandshr) [reply tests/evaluation/retinal_evaluation.py]: Instead of integration test, perhaps we can treat this as a performance monitor that should be done for any updated version changes

This is where this data is
https://www.dropbox.com/sh/mxocvcdy4747esf/AACUBqg830Q3BA7vfOFrMGAQa?dl=0

Comment (calvinytong) [reply tests/evaluation/retinal_evaluation.py]: I'm not a big fan of running two different builds because seems like more overhead than we really want. We can probably get around the data issue if we create a Docker container with the datasets already baked in and push that to quay or some such. If we are just treating this as a performance monitor on each new version, it might just be okay to publish a notebook (whose code would be very similar on each version) when we bump the version number. I think we'd end up with way too many figures that all say about the same thing (not to mention it would be a pain to figure out what to do if we change the figures) if we run too often. We'd have to write even more stuff to deal with those issues as indexing them in some way would be important.

File: tests/evaluation/ej49_geometry1.txt [hunk 5]
```diff
@@ -0,0 +1,49 @@
+-585.0 270.0
```
Comment (edublancas) [inline tests/evaluation/ej49_geometry1.txt]: why there are many geometry files? is that for testing different geometries?

Comment (hooshmandshr) [reply tests/evaluation/ej49_geometry1.txt]: There are 4 geometry files for 4 different 5 minute data sets.

File: tests/evaluation/retinal_evaluation.py (hunk 4 above)
Comment (hooshmandshr) [reply tests/evaluation/retinal_evaluation.py]: Eduardo actually realized that we probably shouldn't make this data public anyway.
We can probably move this code to evaluation to allow for creating stability metric on a single dataset end to end for yass

Then a notebook for retinal data. Any time someone is submitting a new branch they should run this notebook and report the results for people to review perhaps?

Comment (hooshmandshr) [reply tests/evaluation/retinal_evaluation.py]: Agreed.

File: tests/evaluation/retinal_evaluation.py [hunk 6]
```diff
@@ -0,0 +1,282 @@
+import numpy as np
+import matplotlib.pyplot as plt
+import os
+import scipy.io
+import yaml
+
+from numpy import genfromtxt
+from yass.command_line import _run_pipeline
```
Comment (edublancas) [inline tests/evaluation/retinal_evaluation.py]: can you update to use the new function? it's already on master yass/pipeline.py

File: tests/evaluation/retinal_evaluation.py [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+        bin_file = 'ej49_data{}.bin'.format(data_number)
+        geom_file = 'ej49_geometry{}.txt'.format(data_number)
+        config['data']['recordings'] = bin_file
+        config['data']['geometry'] = geom_file
+        with open('config_ej49.yaml', 'w') as yaml_file:
+            yaml.dump(config, yaml_file, default_flow_style=False)
+        # TODO(hooshmand): Replace with the finalized pipeline function.
+        yass_spike_train_file = 'yass_spike_train_{}.csv'.format(data_number)
+        _run_pipeline(
+          config='config_ej49.yaml', output_file=yass_spike_train_file)
```
Comment (edublancas) [inline tests/evaluation/retinal_evaluation.py]: the new function also works with dictionaries, please update

**Comment by calvinytong** [comment]
Found it [here](https://docs.travis-ci.com/user/customizing-the-build/#Build-Timeouts).

10 min without a log message
50 minutes total

File: tests/evaluation/retinal_evaluation.py (hunk 6 above)
Comment (hooshmandshr) [reply tests/evaluation/retinal_evaluation.py]: Cool, one question though
How to indicate what the output spike train file name is?

Comment (edublancas) [reply tests/evaluation/retinal_evaluation.py]: it is always set to spike_train.npy - do you need that as a parameter?

Comment (hooshmandshr) [reply tests/evaluation/retinal_evaluation.py]: I can use os.rename(), would it be ok?

Comment (hooshmandshr) [reply tests/evaluation/retinal_evaluation.py]: I had a request, can you actually modify pipeline_run() to return the spike_train as a numpy.ndarray of shape (N, 2) N being number of spikes?

File: tests/evaluation/retinal_evaluation.py [hunk 8]
```diff
@@ -0,0 +1,268 @@
...
+    acc_plot = EvaluationPlot('EJ Retinal', k_tot_data)
+
+    for data_idx, data_number in enumerate(range(1, k_tot_data + 1)):
+        # Setting up config file for yass.
+        bin_file = 'ej49_data{}.bin'.format(data_number)
+        geom_file = 'ej49_geometry{}.txt'.format(data_number)
+        config['data']['recordings'] = bin_file
+        config['data']['geometry'] = geom_file
+        with open('config_ej49.yaml', 'w') as yaml_file:
+            yaml.dump(config, yaml_file, default_flow_style=False)
```
Comment (edublancas) [inline tests/evaluation/retinal_evaluation.py]: since you are now passing a dictionary, do we still need this yaml.dump?

Comment (hooshmandshr) [reply tests/evaluation/retinal_evaluation.py]: Sorry, took care of it.
